full-node/sov-stf-runner/src/lib.rs

```diff
@@ -147,13 +187,30 @@ where
 
             let mut data_to_commit = SlotCommit::new(filtered_block.clone());
 
-            let slot_result = self
-                .app
-                .apply_slot(Default::default(), &filtered_block, &mut blobs);
+            let slot_result = self.app.apply_slot(
+                Default::default(),
+                filtered_block.header(),
+                &validity_condition,
+                &mut blobs,
+            );
             for receipt in slot_result.batch_receipts {
                 data_to_commit.add_batch(receipt);
             }
-            let next_state_root = slot_result.state_root;
+            let next_state_root: ST::StateRoot = slot_result.state_root;
+            if let Some((prover_instance, guest)) = self.prover.as_mut() {
+                let (inclusion_proof, completeness_proof) = self
```

Comment:
The `run` method overlaps with the code here: https://github.com/Sovereign-Labs/sovereign-sdk/pull/755/files#diff-1f608de90abe33cd820bb3380f66ab5b75a2817443baeed7bc11c3ad21314869R125.

There is:
1. For loop over heights.
2. `da_service.extract_relevant_txs
3. `get_extraction_proof(&filtered_block, &blobs)` but here this is optional.
etc..

So, the `shape` of the code between `demo-prover/host/main` and this `run` function is very similar.
It looks like the only difference is that line 200 here is optional + in 206 we do `                prover_instance.write_to_guest(&next_state_root)` while in hos we do `host.write_to_guest(prev_state_root);` is that ok? 

It would be great if we could make the `sov-stf-runner` generic enough to be used in `demo-prover/host/main` as well. This way the `Runner` could be used in both `host` and `rollup main`

The one approach which comes to my mind is to make the `prober_instance` a trait and insert `rover_instance.write_to_guest` in all the places that is required by host

Reply:
Another way would be to extract the logic in the `for loop` and reuse it here, and in  `demo-prover/host/main` + the `demo-prover` would add the proving step at the end.